src/Workspaces/Core/Portable/Workspace/Solution/Solution.CompilationTracker.cs

```diff
@@ -794,6 +788,58 @@ public IEnumerable<SyntaxTree> GetSyntaxTreesWithNameFromDeclarationOnlyCompilat
                 return clone.GetSymbolsWithName(predicate, filter, cancellationToken).SelectMany(s => s.DeclaringSyntaxReferences.Select(r => r.SyntaxTree));
             }
 
+            public Task<bool> HasCompleteReferencesAsync(Solution solution, CancellationToken cancellationToken)
+            {
+                var state = this.ReadState();
+
+                if (state.HasCompleteReferences.HasValue)
+                {
+                    return Task.FromResult(state.HasCompleteReferences.Value);
+                }
+                else
+                {
+                    return GetOrBuildCompilationInfoAsync(solution, lockGate: true, cancellationToken: cancellationToken)
+                        .ContinueWith(t => t.Result.HasCompleteReferences, cancellationToken, TaskContinuationOptions.ExecuteSynchronously | TaskContinuationOptions.OnlyOnRanToCompletion, TaskScheduler.Default);
+                }
+            }
+
+            private AsyncLazy<bool> _lazyDependentCompleteReference;
+
+            public async Task<bool> HasDependentCompleteReferencesAsync(Solution solution, CancellationToken cancellationToken)
+            {
+                if (_lazyDependentCompleteReference == null)
+                {
+                    // note: solution is captured here, but it will go away once GetValueAsync executes.
+                    Interlocked.CompareExchange(ref _lazyDependentCompleteReference, new AsyncLazy<bool>(c => ComputeDependentCompleteReferenceAsync(solution, c), cacheResult: true), null);
+                }
+
+                return await _lazyDependentCompleteReference.GetValueAsync(cancellationToken).ConfigureAwait(false);
+            }
+
+            public async Task<bool> ComputeDependentCompleteReferenceAsync(Solution solution, CancellationToken cancellationToken)
```

Comment:
I can't for the life of me explain the difference between this method and the previous one. Accident that they are both public?